src/models/state/mempool.rs

```diff
@@ -283,13 +315,22 @@ impl Mempool {
     }
 
     /// Remove from the mempool all transactions that become invalid because
-    /// of this newly mined block. Also update all mutator set data for monitored
-    /// transactions that were not removed in the previous step.
+    /// of a newly received block. Also update all mutator set data for mempool
+    /// transactions that were not removed.
     pub async fn update_with_block(
         &mut self,
         previous_mutator_set_accumulator: MutatorSetAccumulator,
         block: &Block,
     ) {
+        // If we discover a reorganization, we currently just clear the mempool,
+        // as we don't have the ability to roll transaction removal record integrity
```

Comment:
Is is it expected/planned that we will eventually be able to rollback transaction removal record integrity proofs?  ie, is it "just" an engineering issue or a theoretical one?

I think that forgetting user's transactions during re-org could be a pretty serious drawback for neptune, so I'm hoping this can be changed by launch, or at least that we'll be able to tell users it is "coming soon".

Reply:
> Is is it expected/planned that we will eventually be able to rollback transaction removal record integrity proofs? ie, is it "just" an engineering issue or a theoretical one?

This limitation can certainly be solved through "relatively" simple engineering. The easiest way to solve it would be for each node to store multiple copies of the same transaction, with a copy for the last $n$ blocks. So if a reorganization happens, we can take a copy of the transaction at the latest block that both branches agree on and start from there.

This, however, does not solve the problem of adding to the mempool transactions that were mined on the abandoned fork and now would need to be re-added to the mempool.

Consider a scenario where tip changes from `3a` to `3b`
```text
                                      +-- block 2a <-- block 3a     
                                      |                          
 Genesis <-- block 1 <--+                            
                                      |                           
                                     +-- block 2b <-- block 3b
```

The transactions in the mempool at the time of `3a` can be updated to be valid for `3b` provided the client knows a version of the transaction that was valid at block `1`. The transactions that are no longer mined since block `2a` and `3a` were abandoned would also have to be stored by the client in a version valid at block `1`. So it's certainly possible to preserve the mempool after a reorganization that's only a few blocks deep, but you pay a quite heavy price in terms of RAM consumption for this.

Reply:
> This, however, does not solve the problem of adding to the mempool transactions that were mined on the abandoned fork and now would need to be re-added to the mempool

Good point.  

> So it's certainly possible to preserve the mempool after a reorganization that's only a few blocks deep, but you pay a quite heavy price in terms of RAM consumption for this

My thought would be that whenever we remove utxo(s) from the mempool because they were confirmed in a block, we could write them to disk, indexed by block hash.  We could perhaps keep such mempool-blocks to a depth of 100 (or some configurable amount).   We would only need to consult the on-disk store during a re-org and would not be using any additional RAM.   And only re-orgs deeper than 100 would lose tx, which hopefully never happen anyway.

Reply:
Currently the mempool *only* lives in memory and is never persisted on disk. I'm pretty happy with that quality but I guess it's not set in stone. Feel free to open an issue about the limited reorganization capabilities of the mempool.

Reply:
yeah, so I would look at it as the mempool remains unchanged, with everything in mem, but we add a new thing, which might be called mempool-archive, that would be stored on disk, only for re-org usage.

ok, I will open an issue for it to remind us.